src/main/java/graphql/annotations/dataFetchers/MethodDataFetcher.java

```diff
@@ -118,40 +132,67 @@ private Object[] invocationArgs(DataFetchingEnvironment environment, ProcessingE
+                if (parameters[0].getType().isAssignableFrom(Optional.class)) {
+                    return constructNewInstance(constructor, arg);
+                } else {
+                    return constructNewInstance(constructor, optionalArg.orElse(null));
+                }
             } else {
                 List<Object> objects = new ArrayList<>();
-                Map map = (Map) arg;
+                Map map = (Map) optionalArg.orElseGet(Collections::emptyMap);
                 for (Parameter parameter : parameters) {
                     String name = toGraphqlName(parameter.getAnnotation(GraphQLName.class) != null ? parameter.getAnnotation(GraphQLName.class).value() : parameter.getName());
-                    objects.add(buildArg(parameter.getParameterizedType(), ((GraphQLInputObjectType) graphQLType).getField(name).getType(), map.get(name)));
+                    if (!map.containsKey(name)) {
+                        objects.add(null);
+                    } else {
+                        objects.add(buildArg(parameter.getParameterizedType(), ((GraphQLInputObjectType) graphQLType).getField(name).getType(), map.get(name)));
+                    }
                 }
                 return constructNewInstance(constructor, objects.toArray(new Object[objects.size()]));
             }
         } else if (p instanceof ParameterizedType && graphQLType instanceof GraphQLList) {
-            List<Object> list = new ArrayList<>();
-            Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
-            GraphQLType wrappedType = ((GraphQLList) graphQLType).getWrappedType();
+            if (((ParameterizedType) p).getRawType() == Optional.class) {
+                if (!optionalArg.isPresent()) {
+                    return null;
+                } else {
+                    Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
+                    return Optional.ofNullable(buildArg(subType, graphQLType, arg));
+                }
+            } else {
+                List<Object> list = new ArrayList<>();
+                Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
+                GraphQLType wrappedType = ((GraphQLList) graphQLType).getWrappedType();
 
-            for (Object item : ((List) arg)) {
-                list.add(buildArg(subType, wrappedType, item));
+                for (Object item : ((List) optionalArg.orElseGet(Collections::emptyList))) {
+                    list.add(buildArg(subType, wrappedType, item));
+                }
+                return list;
+            }
+        } else if (p instanceof ParameterizedType && ((ParameterizedType) p).getRawType() == Optional.class) {
+            Type subType = ((ParameterizedType) p).getActualTypeArguments()[0];
+            if (arg == null) {
+                return null;
+            } else {
+                return Optional.ofNullable(buildArg(subType, new GraphQLUndefined(), arg));
```

Comment:
Here I think we have a major issue : the second parameter should be graphqlType, we are just creating an optional of the arg parameter. If subtype is another parameterized type (a list), it won't be correctly parsed. Anyway it's strange to use GraphqlUndefined here since the value is defined.